File: driver-intro/code/rtc.py
import struct
import fcntl
import ioctl

# struct rtc_time {
# 	int tm_sec;
# 	int tm_min;
# 	int tm_hour;
# 	int tm_mday;
# 	int tm_mon;
# 	int tm_year;
# 	int tm_wday;
# 	int tm_yday;
# 	int tm_isdst;
# };

fmt = "iiiiiiiii"
time = bytearray(struct.calcsize(fmt))

fp = open("/dev/rtc0")
fcntl.ioctl(fp, ioctl.RTC_RD_TIME, time)
sec, min, hour, mday, mon, year, wday, yday, isdst = struct.unpack(fmt, time)
fp.close()

print("Time: {0:02}:{1:02}:{2:02}".format(hour, min, sec))
print("Date: {0}/{1}/{2}".format(mday, mon+1, year+1900))

File: driver-intro/code/rtc.c
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/rtc.h>

#include <error.h>
#include <errno.h>

int main()
{
	int fd;
	struct rtc_time rtc_tm;
	int ret;

	fd = open("/dev/rtc0", O_RDONLY);
	if (fd == -1) {
		error(1, errno, "error opening RTC");
	}

	ret = ioctl(fd, RTC_RD_TIME, &rtc_tm);
	if (ret == -1) {
		error(1, errno, "error reading time");
	}

	printf("Time: %02d:%02d:%02d\n", rtc_tm.tm_hour, 
	       rtc_tm.tm_min, rtc_tm.tm_sec);
	
	return 0;
}
